### DISCUSSION TITLE
Best way to use @urql/exchange-auth

### DISCUSSION THREAD

**Discussion by WildEgo** [urql-graphql/urql#2089] [Q&A] [answered]
Hey guys!

Pretty new to this package and I found that it has a way to handle authentication with OAuth in an official package, so I'm wondering what is the best way to handle it? I know I need a way to access information on my current user and also need to check if the current user is there or not in the layouts, I've done it in the past with Apollo but in a way I'd think was not the best, so I'm currently searching for tips or solutions with examples that I can read!

Kind regards,
Luis Bizarro.

**Comment by JoviDeCroock** [comment, marked as answer]
There's an example [ìn our repo](https://github.com/FormidableLabs/urql/blob/main/examples/with-refresh-auth/src/client.js#L34) and [the documentation sheds light as well](https://formidable.com/open-source/urql/docs/advanced/authentication/)

It's hard to give concrete advice without information about the case it self 😅

**Reply by WildEgo** [reply]
One of the questions would be how do I login, for example how do I trigger an event of some sort etc and how can I check if the token is valid from the layout's side so I know what to display and what not, I know it's probably a dumb question but from my background I've worked with Apollo and wrote a full provider that did everything by default

**Reply by JoviDeCroock** [reply]
so what I've done on previous projects is I have a provider tracking the authenticated-state at the root of the project, when the user logs in a token will be set in `localStorage` (for the purpose of example). Now with `getAuth` we'll get that token and `refresh` it if it errors out. 

Which is all in the example btw

- [state tracking](https://github.com/FormidableLabs/urql/blob/main/examples/with-refresh-auth/src/App.jsx#L11)
- [login success](https://github.com/FormidableLabs/urql/blob/f0d2d16644ef663e7c8b52a89f511b512d7b73bc/examples/with-refresh-auth/src/authStore.js#L4)
- [in exchange getting an initial token state](https://github.com/FormidableLabs/urql/blob/main/examples/with-refresh-auth/src/client.js#L36)
- [refresh if needed](https://github.com/FormidableLabs/urql/blob/main/examples/with-refresh-auth/src/client.js#L46-L54)
- [check if we face an auth error, if we do we dispose auth-state and call getAuth](https://github.com/FormidableLabs/urql/blob/main/examples/with-refresh-auth/src/client.js#L86-L88)

Ofcourse this example is a bit naive in the sense that there are possibly unauthenticated vs authenticated queries but this lines out the gist of it very well imo. This in combination with the example from the docs

**Reply by WildEgo** [reply]
Oh thank you I really hadn't found that example, I don't think it was linked anywhere besides ur reply really, I kinda just auto piloted the reply cuz it had a link for the documentation and it had got me kinda lost, I'll read it and I think that's about it for my question, thanks mate!

**Reply by JoviDeCroock** [reply]
No problem, if that doesn't help out, do please get back to me with the stuff that's unclear that way we can improve our examples/documentation

**Reply by WildEgo** [reply]
Hey mate I've done this, now I'm facing a problem I haven't found any documentation on how to get a typed schema working, are there any examples I haven't seen?

**Reply by JoviDeCroock** [reply]
Very vague question but assuming you mean for graphcache? https://github.com/JoviDeCroock/graphcache-code-generation/tree/main/example

**Reply by WildEgo** [reply]
If it works kinda like Apollo does, I don't mind having to run a command to get typescript data, I just need to have my types generated so my data isn't identified as ANY

**Reply by JoviDeCroock** [reply]
We advise this tool always https://www.graphql-code-generator.com/ which can be used in the generics of our hooks
